### PR TITLE
Add configuration api http reference

### PR DESCRIPTION
**Pull request by yaron2**
Closes https://github.com/dapr/docs/issues/2493.

### CODE REVIEW

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
+---
+type: docs
+title: "Configuration API reference"
+linkTitle: "Configuration API"
+description: "Detailed documentation on the configuration API"
+weight: 700
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
weight: 650
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+---
+
+## Component file
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:9]: You can remove this component file section. It is not needed here in the API spec.

Comment (yaron2) [inline daprdocs/content/en/reference/api/configuration_api.md]: Done

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+  version: v1
+  metadata:
+  - name:<KEY>
+    value:<VALUE>
+```
+
+| Setting | Description                                                                            |
+| ------- |----------------------------------------------------------------------------------------|
+| `metadata.name` | The name of the configuration store.                                                   |
+| `spec.metadata` | An open key value pair metadata that allows a binding to define connection properties. |
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]: This is a strange description
```suggestion
| `spec.metadata` | key/value pair metadata that is configured for each configuration store type. |
```

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: This is not relevant anymore as this section was removed.

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+## Get Configuration
+
+This endpoint lets you get configuration from a store.
+
+### HTTP Request
+
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:14]:
```suggestion
Return all configuration items from the named store.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`daprPort` | The Dapr port
+`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.
+
+#### Query Parameters
+
+If no query parameters are provided, all configuration items will be returned.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
If no query parameters are provided, all configuration items are returned.
```

Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]: Then Add this example to show retrieving all the items
```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/mystore
```

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Added

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+To specifiy the keys of the configuration items to get, use one or more `key` query parameters. For example:
+
+```
+GET http://localhost:3500/v1.0/configuration/mystore?key=config1&key=config2
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
GET http://localhost:3500/v1.0-alpha1/configuration/mystore?key=config1&key=config2
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+`500`  | Failed to get configuration
+
+#### Response Body
+
+JSON-encoded value
+
+### Example
+
+```shell
+curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/example-config?key=myConfigKey'
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore?key=myConfigKey' 
```

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 6 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+> The above command returns the following JSON:
+
+```json
+[{"key":"myConfigKey","value":"myConfigValue"}]
+```
+
+## Subscribe Configuration
+
+This endpoint lets you subscribe to configuration changes.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
This endpoint lets you subscribe to configuration changes. Notifications happen when values are updated or deleted in the configuration store. This enables the application to react to configuration changes.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+### HTTP Request
+
+```
+GET http://localhost:<daprPort>/v1.0/configuration/<storename>/subscribe
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>/subscribe
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`daprPort` | The Dapr port
+`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.
+
+#### Query Parameters
+
+If no query parameters are provided, all configuration items will subscribed to.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
If no query parameters are provided, all configuration items are subscribed to.
```

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 7 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 8 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Added

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 9 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 10 above)
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]: Then Add this example to show subscribing to all configuration item updates
```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/mystore/subscribe
```

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+To specifiy the keys of the configuration items to subscribe to, use one or more `key` query parameters. For example:
+
+```
+GET http://localhost:3500/v1.0/configuration/mystore/subscribe?key=config1&key=config2
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
GET http://localhost:3500/v1.0-alpha1/configuration/mystore/subscribe?key=config1&key=config2
```

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 10 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Added

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 11 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+`500`  | Failed to subscribe to configuration changes
+
+#### Response Body
+
+JSON-encoded value
+
+### Example
+
+```shell
+curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/subscribe?key=myConfigKey'
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore/subscribe?key=myConfigKey' 
```

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+Code | Description
+---- | -----------
+`200`  | Get operation successful
+`400`  | Configuration store is missing or misconfigured or malformed request
+`500`  | Failed to get configuration
+
+#### Response Body
+
+JSON-encoded value
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
JSON-encoded value of key/value pairs for each configuration item.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+> The above command returns the following JSON:
+
+```json
+{
+  "id": "<unique-id>"
+}
+```
+
+The returned `id` parameter can be used to unsubscribe to the keys provided on the API call.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
The returned `id` parameter can be used to unsubscribe to the specific set of keys provided on the subscribe API call. This should be retained by the application.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+## Unsubscribe Configuration
+
+This endpoint lets you unsubscribe to configuration changes.
+
+### HTTP Request
+
+```
+GET http://localhost:<daprPort>/v1.0/configuration/<subscription-id>/unsubscribe
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]: Do you need store name here?
```suggestion
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>/<subscription-id>/unsubscribe
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+`500`  | Failed to unsubscribe to configuration changes
+
+#### Response Body
+
+None
+
+### Example
+
+```shell
+curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe'
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:179]: Do you need store name here
```suggestion
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+```shell
+curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
+```
+
+## Optional Application (User Code) Routes
+
+### Provide a route for Dapr to send configuration changes
+
+When subscribing to configuration changes, Dapr will invoke the application whenever a configuration item changes.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
When subscribing to configuration changes, Dapr can invoke an endpoint on the application whenever a configuration item changes.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+
+None
+
+### Example
+
+```shell
+curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
+```
+
+## Optional Application (User Code) Routes
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
## Optional application (user code) routes
```

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 15 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: No

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 16 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md:179]: No

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 18 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 17 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`appPort` | The application port
+`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above
+`key` | The key of the changed configuration item
+
+#### Request Body
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:202]: What is the example showing here? is this supposed to show the routes on the application?

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 14 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 13 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 19 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md:202]: Yes, we do the exact same thing on the pub/sub API reference.

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+```
+GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>
+```
+
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`daprPort` | The Dapr port
+`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+```
+GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>/subscribe
+```
+
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`daprPort` | The Dapr port
+`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+```
+POST http://localhost:<appPort>/configuration/<store-name>/<key>
+```
+
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`appPort` | The application port
+`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+        "value": "abcdefgh",
+        "version": "1.1"
+    ]
+}
+```
+
+
+## Next Steps
+
+- [Configuration api overview]({{< ref configuration-api-overview.md >}})
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
- [Configuration API overview]({{< ref configuration-api-overview.md >}})
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+- [How-To: Manage configuration]({{< ref howto-manage-configuration.md >}})
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
- [How-To: Manage configuration from a store]({{< ref howto-manage-configuration.md >}})
```

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 23 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 22 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 21 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 20 above)
Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md]: Changed

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 19 above)
Comment (msfussell) [reply daprdocs/content/en/reference/api/configuration_api.md:202]: If you are referring to this https://docs.dapr.io/reference/api/pubsub_api/#http-response-body then that is not what is shown below. What is `version`. And I do not understand how is this used? Whereas subscription for pub/sub returns a list of the routes, from the app, what is this calling on the app? Please explain.

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md:202]: No, it's referring to https://docs.dapr.io/reference/api/pubsub_api/#provide-routes-for-dapr-to-deliver-topic-events.

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+
+```shell
+curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
+```
+
+## Optional application (user code) routes
+
+### Provide a route for Dapr to send configuration changes
+
+When subscribing to configuration changes, Dapr invokes the application whenever a configuration item changes.
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:186]: Your application can have a `/configuration` endpoint that is called for all key updates that are subscribed to. The endpoint(s) can be made more specific for a given configuration store by adding `/<store-name>` and for a specific key by adding  `/<store-name>/<key>` to the route.

Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md]:
```suggestion
When subscribing to configuration changes, Dapr invokes the application whenever a configuration item changes. Your application can have a `/configuration` endpoint that is called for all key updates that are subscribed to. The endpoint(s) can be made more specific for a given configuration store by adding /<store-name> and for a specific key by adding /<store-name>/<key> to the route.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`appPort` | The application port
+`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
+
+#### Request Body
+
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:203]:
```suggestion
A list of configuration items for a given subscription id. Configuration items can have a version associated with them, which is returned in the notification.
```

File: daprdocs/content/en/reference/api/configuration_api.md [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+```
+POST http://localhost:<appPort>/configuration/<store-name>/<key>
+```
+
+#### URL Parameters
+
+Parameter | Description
+--------- | -----------
+`appPort` | The application port
+`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```
Comment (msfussell) [inline daprdocs/content/en/reference/api/configuration_api.md:199]:
```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
`key` | The key subscribed to
```

File: daprdocs/content/en/reference/api/configuration_api.md (hunk 19 above)
Comment (msfussell) [reply daprdocs/content/en/reference/api/configuration_api.md:202]: Question: Like pub/sub are there any success/failure codes that the application can return to Dapr?

Comment (yaron2) [reply daprdocs/content/en/reference/api/configuration_api.md:202]: No

**Review by msfussell — APPROVED** [review]
Ta Dah!

**Pull request merged**